### PR TITLE
runtime: register atfork handler to re-initialize internal flangrti locks at fork

### PR DESCRIPTION
**Pull request by pawosm-arm**
The llcrit.c code of flangrti runtime library creates internal locks
which are initialized on the first encounter of any Fortran procedure
that contains at least one OpenMP directive. The flang compiler emits
a call to one of the two functions modified by this patch at the
beginning of every Fortran procedure containing at least one OpenMP
directive. When executed for the first time, these functions
initialize internal locks and modify corresponding static global
variable confirming the initialization has actually occured. On any
subsequent execution, these functions make use of those locks.

Unfortunately, the values of those global indicator variables are not
being reset at fork, hence the locks are not being initialized in the
child processes. Any attempt of using those internal locks results in
a runtime crash of the child process.

This patch registers atfork handler executed as the first thing by any
of the children process.

A test case depicting the problem is also provided with this patch.

### CODE REVIEW

**Comment by SouraVX** [comment]
Changes, look legit. But when I tried the test case(part of this PR) without code changes:
```
           12
            0
            7
           15
            8
            2
            3
            1
           10
            4
           11
            6
            9
           13
           14
            5
FAIL
```
>Unfortunately, the values of those global indicator variables are not
being reset at fork, hence the locks are not being initialized in the
child processes. Any attempt of using those internal locks results in
a runtime crash of the child process.

Based on your comment, I was expecting a CRASH :). Am I missing something, or could you provide a repro for crash.

**Comment by SouraVX** [comment]
Steps followed while building:
```
clang -c -I/home/flang/test/mp_correct/lit/../src  -fopenmp  /home/flang/test/mp_correct/lit/../src/fork.c
clang -c -I/home/flang/test/mp_correct/lit/../src  -fopenmp  /home/flang/test/mp_correct/lit/../src/check.c -o check.o
flang -c -I/home/flang/test/mp_correct/lit/../src  -fopenmp  /home/flang/test/mp_correct/lit/../src/fork_omp.F90
flang   -fopenmp fork_omp.o fork.o check.o -o a.out
```

**Comment by pawosm-arm** [comment]
If you see `FAIL`, then there was a crash and it was detected by waitpid(). If you really want to post-mortem-debug it, enable core dumps in your system (with `ulimit -c unlimited`), in effect, whenever you see `FAIL` message a core file should appear, which then you can open in a debugger.
When a child process stops or terminates, `SIGCHLD` is sent to the parent process. The default response to the signal is to ignore it. Therefore if you want this crash to be more visible, you'd have to un-ignore SIGCHLD signal. I think checking the status set by the `waitpit()` function and drop a failure message seems neater.

**Comment by pawosm-arm** [comment]
@SouraVX are you OK with this patch?

**Review by kiranchandramohan — APPROVED** [review]
LGTM. Please wait for an approval from @SouraVX or @shivaramaarao.

**Comment by SouraVX** [comment]
I noted following behavior:
**Before patch (Signal received by the process):**
```
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_KILLED, si_pid=16277, si_uid=1001, si_status=SIGSEGV, si_utime=1, si_stime=0} 
```
**After patch (Signal received by the process):**
```
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16696, si_uid=1001, si_status=0, si_utime=16, si_stime=2} ---
```
If that's the intended thing, I'm happy :)
Some tests are still running, is this Okay to wait for 1 more day ? 
Thanks a lot!

**Comment by pawosm-arm** [comment]
Sure it can wait. Thanks for looking at it @SouraVX

**Comment by SouraVX** [comment]
BTW @bryanpkc did you get a chance to have a look ?

**Review by SouraVX — APPROVED** [review]
LGTM. Thanks!
Please wait for approval from @bryanpkc  also.

**Comment by kiranchandramohan** [comment]
@SouraVX just want to highlight that technically waiting for @bryanpkc is not required as per the Pull Request process. The PR process only requires that two reviewers test and approve the PR on Aarch64 and AMD. If that is not everyone's understanding then we can discuss the PR processing again.
However it is a good practise to wait for @bryanpkc.

**Comment by SouraVX** [comment]
> @SouraVX just want to highlight that technically waiting for @bryanpkc is not required as per the Pull Request process. The PR process only requires that two reviewers test and approve the PR on Aarch64 and X86_64. If that is not everyone's understanding then we can discuss the PR processing again.
> However it is a good practise to wait for @bryanpkc (or all organizations involved).

Thanks for clarification. No issues from our side :)

File: runtime/flangrti/llcrit.c
Comment (bryanpkc) [inline runtime/flangrti/llcrit.c:72]: Nit: Use PGI coding style?

**Review by bryanpkc — APPROVED** [review]
LGTM. Built fine in our code base.

**Pull request merged**
